---
title: "Phyloseq workflow"
author: "Alexis Carteron & Simon Morvan"
date: "9 mars 2018"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r , include=TRUE}

# Analysis of microbiological communities

## Load packages ####
library(phyloseq)
library(vegan)
library(ggplot2)

#### Data import and processing ####
## Sample data
soiltab <- read.table('data/soil.csv', sep = ';', row.names = 1, header = TRUE)

# Taxonomy table
load("data/taxotab.01.05.rdata")

# ASV table
load("data/seqtab.nochim.rdata")

# Construct phyloseq object (straightforward from dada2 outputs)
ps <- phyloseq(otu_table(seqtab.nochim, taxa_are_rows=FALSE), 
               sample_data(soiltab), 
               tax_table(as.matrix(taxo$tax)))
ps

# Reorder factors
sample_data(ps)$horizon = factor(sample_data(ps)$horizon, levels = c('L', 'F', 'H', 'Ae', 'B'))
sample_data(ps)$forest = factor(sample_data(ps)$forest, levels = c('AS', 'Mixed', 'FG'))

## Start directly from phyloseq object
#load(file = "data/ps.rdata")

#### Filtering and transformation ####
ntaxa(ps)

## Subset of the dataset with only fungi
ps.fungi = subset_taxa(ps, Kingdom=="k__Fungi")
ntaxa(ps.fungi)

## Remove singletons and doubletons
ps.fungi.nosd = filter_taxa(ps.fungi, function(x) sum(x) > 2, TRUE)
ntaxa(ps.fungi.nosd)

## Shifted log transformation
ps.fungi.nosd.log <- transform_sample_counts(ps.fungi.nosd, function(x) log(x+1))

## Agglomerate taxa at the species level
ps.fungi.nosd.log.sp = tax_glom(ps.fungi.nosd.log, "Species", NArm = FALSE) # NArm = FALSE because it keeps ESV not taxonomically assigned at the species level

## Other transformation
## e.g. Hellinger transformation
ps.fungi.nosd.hel = transform_sample_counts(ps.fungi.nosd, function(x) sqrt(x/sum(x)))

#### Exploration ####
## Details on the object
ntaxa(ps.fungi.nosd.log)
nsamples(ps.fungi.nosd.log)
rank_names(ps.fungi.nosd.log)
sample_variables(ps.fungi.nosd.log)
otu_table(ps.fungi.nosd.log)[1:2, 1:2]
tax_table(ps.fungi.nosd.log)[1:5, 1:6]

# Sequencing depth per samples
plot_bar(ps, x = "Sample", y = "Abundance", fill = "Kingdom") +
  geom_bar(stat="identity", color = "black") +
  labs(x = "Sample ID", y = "Reads abundance")

plot_bar(ps.fungi,  x = "Sample", y = "Abundance", fill = "Kingdom") +
  geom_bar(stat="identity", color = "black") +
  labs(x = "Sample ID", y = "Reads abundance of fungi")

plot_bar(ps.fungi,  x = "Sample", y = "Abundance", fill = "Phylum") +
  geom_bar(stat="identity") +
  labs(x = "Sample ID", y = "Reads abundance of fungi")

plot_bar(ps.fungi, fill = "Phylum") +
  geom_bar(stat="identity") +
  facet_wrap(horizon~forest, ncol = 3, scales = "free_x") +
  labs(x = "Sample ID", y = "Reads abundance of fungi")

#### Ordination ####
myco_names <- c(`AS` = "Maple plots", `Mixed` = "Mixed plots", `FG` = "Beech plots")

## nMDS with Bray-Curtis dissimilarity
ord.nmds.bray <- ordinate(ps.fungi.nosd.log.sp, method="NMDS", k = 2, try = 100, distance = "bray")

## Quick plot
plot_ordination(ps.fungi.nosd.log.sp, ord.nmds.bray, type="samples", color = "horizon")

## Full plot
plot_ordination(ps.fungi.nosd.log.sp, ord.nmds.bray, type="samples", color = "horizon") +
  geom_point(size = 8, alpha = .5, show.legend = TRUE) +
  scale_color_manual(name = "Horizon", values = c('darkgreen', 'sienna4', 'grey1', 'grey60', 'darkorange2')) +
  labs(x = "Axis 1", y = "Axis 2", caption = paste("Stress =", round(ord.nmds.bray$stress, 2))) +
  facet_wrap(~forest) +
  stat_ellipse(type = "t", level = 0.8) + # 80% confidence interval ellispes
  coord_fixed() +
  theme(text = element_text(size=25), strip.text = element_text(size=25))

## nMDS with Sorensen dissimilarity
ord.nmds.bin <- ordinate(ps.fungi.nosd.log.sp, method="NMDS", k = 2, try = 100, distance = "bray", binary =TRUE)

plot_ordination(ps.fungi.nosd.log.sp, ord.nmds.bin, type="samples", color = "horizon") +
  geom_point(size = 8, alpha = .5, show.legend = TRUE) +
  scale_color_manual(name = "Horizon", values = c('darkgreen', 'sienna4', 'grey1', 'grey60', 'darkorange2')) +
  labs(x = "Axis 1", y = "Axis 2", caption = paste("Stress =", round(ord.nmds.bin$stress, 2))) +
  facet_wrap(~forest) +
  stat_ellipse(type = "t", level = 0.8) + # 80% confidence interval ellispes
  coord_fixed() +
  theme(text = element_text(size=25), strip.text = element_text(size=25))

#### Permanova ####
# Export data
metadata <- as(sample_data(ps.fungi.nosd.log.sp), "data.frame")

## Bray-Curtis measure
## Horizon and forest effect with interaction and block
adonis(phyloseq::distance(physeq = ps.fungi.nosd.log.sp, method="bray") ~ horizon * forest,
      strata = metadata$block,
      data = metadata,
      permutations = 99999)

## Test for the homogeneity of dispersion with Bray-Curtis measures
betadisp <- betadisper(phyloseq::distance(physeq = ps.fungi.nosd.log.sp, method="bray"), metadata$forest)
permutest(betadisp, permutations = 99999) 

betadisp <- betadisper(phyloseq::distance(physeq = ps.fungi.nosd.log.sp, method="bray"), metadata$horizon)
permutest(betadisp, permutations = 99999)

## Sorensen measure
## Horizon and forest effect with interaction and block
adonis(phyloseq::distance(ps.fungi.nosd.log.sp, method="bray", binary = TRUE) ~ horizon * forest, strata = metadata$block,
       data = metadata,
       permutations = 99999)

## Test for the homogeneity of dispersion for Bray data
betadisp <- betadisper(phyloseq::distance(ps.fungi.nosd.log.sp, method="bray", binary = TRUE), metadata$forest)
permutest(betadisp, permutations = 99999) 

betadisp <- betadisper(phyloseq::distance(ps.fungi.nosd.log.sp, method="bray", binary = TRUE), metadata$horizon)
permutest(betadisp, permutations = 99999)
  
#### Redundancy Analysis (RDA) ####
ord.rda <- ordinate(ps.fungi.nosd.log.sp, formula = ps.fungi.nosd.log.sp ~ Carbon+Nitrogen+pH, method="CAP", distance = "bray")

# Get stat
RsquareAdj(ord.rda) 
anova.cca(ord.rda)

# Check variance inflation factors 
vif.cca(ord.rda)

## Quick plot
plot_ordination(ps.fungi.nosd.log.sp, vegan::scores(ord.rda, scaling=1), type="sites",  color = "horizon") 

## Full plot
# Add the environmental variables as arrows
arrowmat = vegan::scores(ord.rda, display = "bp", scaling = 1)
# Add labels, make a data.frame
arrowdf <- data.frame(arrowmat)
rownames(arrowdf) <- c("Carbon","Nitrogen","pH")
arrowdf <- data.frame(labels = c("Carbon","Nitrogen","pH"), arrowmat)

# Get eigenvector values
evals <- ord.rda$CCA$eig/sum(ord.rda$CCA$eig)

# plot
plot_ordination(ps.fungi.nosd.log.sp, vegan::scores(ord.rda, scaling = 1), type="sites",  color = "horizon") + 
  coord_fixed(sqrt(evals[2] / evals[1])) +
  geom_point(size = 8, alpha = .5) +
  scale_color_manual(name = "Horizon", values = c('darkgreen', 'sienna4', 'grey1', 'grey60', 'darkorange2')) +
  geom_segment(aes(xend = CAP1, yend = CAP2, x = 0, y = 0), size = .7, data = arrowdf, color = "black", arrow = arrow(length = unit(0.05, "npc"))) +
  geom_text(aes(x = CAP1*1.15, y = CAP2*1.15, color = NULL, label = labels), size = 6, data = arrowdf) +
  facet_wrap(~forest, labeller = as_labeller(myco_names)) +
  theme(text = element_text(size=18), strip.text = element_text(size=25)) +
  labs(x = paste("Constrained axis 1 (",round(evals[1]*100, 0),"%)"), y = paste("Constrained axis 2 (",round(evals[2]*100, 0),"%)"))
```

